app/models/activity.rb

```diff
@@ -0,0 +1,58 @@
+class Activity < ActiveRecord::Base
+  before_create { |activity| activity.event = generate_key(activity.object, activity.event) }
+
+  belongs_to :object, polymorphic: true
+  belongs_to :actor, inverse_of: :activities, class_name: User.name
+  has_many :course_notifications, inverse_of: :activity, dependent: :destroy
+  has_many :user_notifications, inverse_of: :activity, dependent: :destroy
+
+  USER_NOTIFICATION_TYPES = [:popup, :email]
+  COURSE_NOTIFICATION_TYPES = [:feed]
+
+  # Create notifications based on an activity.
+  #
+  # @param [User, Course] to the user who will receive notifications or the course which will show
+  #   activity feeds.
+  # @param [Array<Symbol>, Symbol] types the formats of the notifications. If 'to' is a user 'types'
+  #   can be :popup, :email or both. If 'to' is a course 'types' can only be :feed.
+  def notify(to, types: nil)
+    types = [*types]
+    user_level_notify(to, types) if to.is_a?(User)
+    course_level_notify(to, types) if to.is_a?(Course)
+  end
+
+  private
+
+  # Generate the key of activity.
+  #
+  # @example generate_key(@course,:created) => courses.notifications.created
+  #
+  # @param trackable The object that will be tracked.
+  # @param [Symbol] event the name of the activity indicating which event happened.
+  # @return [String]
+  def generate_key(trackable, event)
+    "#{trackable.class.name.underscore.pluralize}.notifications.#{event}"
+  end
+
+  # Create user notifications based on an activity.
+  #
+  # @param [User] user the user who will receive notifications
+  # @param [Array<Symbol>] types the formats of the notifications.
+  def user_level_notify(user, types)
+    types.each do |type|
+      fail 'Notification type is invalid' unless USER_NOTIFICATION_TYPES.include?(type)
+      UserNotification.create(activity: self, user: user, notification_type: type)
+    end
+  end
+
+  # Create course notifications based on an activity.
+  #
+  # @param [Course] course the course which will show activity feeds.
+  # @param [Array<Symbol>] types the formats of the notifications.
+  def course_level_notify(course, types)
+    types.each do |type|
+      fail 'Notification type is invalid' unless COURSE_NOTIFICATION_TYPES.include?(type)
+      Course::Notification.create(activity: self, course: course, notification_type: type)
```

Comment:
why is this eager? Shouldn't you build them all and only execute the notification when the activity is saved?